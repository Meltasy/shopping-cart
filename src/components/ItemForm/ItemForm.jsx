import { useState } from 'react'
import styled from 'styled-components'
import Icon from '@mdi/react'
import { mdiBasketPlusOutline } from '@mdi/js'

const StyledForm = styled.form`
  width: 100%;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-end;
  gap: 0.6rem;
  @media (max-width: 480px) {
    gap: 0.3rem;
  }
`

const buttonStyles =`
  font-family: inherit;
  font-size: inherit;
  font-style: inherit;
  color: var(--background-color);
  background-color: var(--primary-color);
  margin: 0.1rem;
  outline: none;
  border: none;
  border-radius: 5px;
  &:hover {
    color: var(--primary-color-dark);
    background-color: var(--primary-color-light);
  }
`

const StyledButton = styled.button`
  ${buttonStyles}
  padding: 0.3rem 0.6rem;
`

const StyledIconButton = styled.button`
  ${buttonStyles}
  padding: 0.5rem 0.6rem 0.3rem 0.5rem;
`

const StyledInput = styled(StyledButton)`
  max-width: 2rem;
`

const ItemForm = ({ itemId, price, onSubmit }) => {
  const [value, setValue] = useState('1')
  const [prevValue, setPrevValue] = useState('1')

  function handleChange(e) {
    setValue((e.target.value))
  }

  function handleDecrease() {
    const num = Number(value) || 0
    if (num > 0) {
      setValue(String(num - 1))
      setPrevValue(String(num - 1))
    }
  }

  function handleIncrease() {
    const num = Number(value) || 0
    if (num < 10) {
      setValue(String(num + 1))
      setPrevValue(String(num + 1))
    }
  }

  function handleAdd(e) {
    e.preventDefault()
    const num = Number(value)
    if (!num || num <= 0) return
    onSubmit(itemId, price, num)
    setValue('0')
    setPrevValue('0')
  }

  return (
    <>
      <StyledForm
        onSubmit={handleAdd}>
      <div>
        <StyledButton type='button' onClick={handleDecrease}>-</StyledButton>
          <StyledInput as='input'
            type='text'
            id='quantity'
            name='quantity'
            min='1'
            max='10'
            pattern='^\d+$'
            value={value}
            onChange={handleChange}
            onFocus={() => {
              setPrevValue(value)
              setValue('')
            }}
            onBlur={() => {
              if (value === '') setValue(prevValue)
            }}
          />
        <StyledButton type='button' onClick={handleIncrease}>+</StyledButton>
      </div>
      <StyledIconButton type='submit'>
        <Icon path={mdiBasketPlusOutline} size={1} />  
      </StyledIconButton>
    </StyledForm>
  </>
  )
}

export default ItemForm
