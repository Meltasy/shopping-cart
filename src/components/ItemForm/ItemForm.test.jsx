import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ItemForm from './ItemForm'

describe('Shop item form component', () => {
  it('renders shop item form', () => {
    const { container } = render(<ItemForm />)
    expect(container).toMatchSnapshot()
  })

  it('renders input for form', () => {
    render(<ItemForm />)
    screen.debug()
    expect(screen.getByRole('textbox'))
  })

  it('renders all three buttons', () => {
    render(<ItemForm />)
    expect(screen.getAllByRole('button').length).toBe(3)
  })

  it('renders all buttons with the correct text', () => {
    render(<ItemForm />)
    expect(screen.getAllByRole('button')[0].textContent).toMatch('-')
    expect(screen.getAllByRole('button')[1].textContent).toMatch('+')
    expect(screen.getAllByRole('button')[2].querySelector('svg')).toBeInTheDocument()
  })

  it('increases quantity by 1 after button click', async () => {
    render(<ItemForm />)
    const user = userEvent.setup()
    const button = screen.getByRole('button', {name: '+'})
    await user.click(button)
    expect(screen.getByDisplayValue('2')).toBeInTheDocument()
  })

  it('decreases quantity by 1 after button click, if quantity > 0', async () => {
    render(<ItemForm />)
    const user = userEvent.setup()
    const increaseBtn = screen.getByRole('button', {name: '+'})
    const decreaseBtn = screen.getByRole('button', {name: '-'})
    await user.tripleClick(increaseBtn)
    await user.click(decreaseBtn)
    expect(screen.getByDisplayValue('3')).toBeInTheDocument()
  })

  it('checks initial value of 1 and updates when user types', async () => {
    render(<ItemForm />)
    const user = userEvent.setup()
    const input = screen.getByRole('textbox')
    expect(input.value).toBe('1')
    await user.type(input, '7')
    expect(input).toHaveValue('7')
  })
})

describe('Shop item form calling onSubmit', () => {
  it('calls onSubmit with correct arguments when form submitted', async () => {
    const mockSubmit = vi.fn()
    render(<ItemForm itemId='123' price={35.50} onSubmit={mockSubmit} />)
    const user = userEvent.setup()
    const input = screen.getByRole('textbox')
    const submitBtn = screen.getAllByRole('button')[2]
    await user.clear(input)
    await user.type(input, '3')
    await user.click(submitBtn)
    expect(mockSubmit).toHaveBeenCalledWith('123', 35.50, 3)
    expect(input).toHaveValue('0')
  })

  it('does not call onSubmit when value is 0', async() => {
    const mockSubmit = vi.fn()
    render(<ItemForm itemId='123' price={35.50} onSubmit={mockSubmit} />)
    const user = userEvent.setup()
    const input = screen.getByRole('textbox')
    const submitBtn = screen.getAllByRole('button')[2]
    await user.clear(input)
    await user.type(input, '0')
    await user.click(submitBtn)
    expect(mockSubmit).not.toHaveBeenCalled()
  })
})
