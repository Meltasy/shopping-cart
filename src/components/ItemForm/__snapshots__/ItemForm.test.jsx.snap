// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop item form component > renders shop item form 1`] = `
<div>
  <form
    class="sc-FEMpB hpBXsI"
  >
    <label
      class="sc-fLDLck heNBuh"
      for="quantity"
    >
      Quantity:
    </label>
    <div>
      <button
        class="sc-gjcoXW eGPGfw"
        type="button"
      >
        -
      </button>
      <input
        class="sc-gjcoXW sc-dVBluf eGPGfw frlVcV"
        id="quantity"
        max="10"
        min="1"
        name="quantity"
        pattern="^\\d+$"
        type="text"
        value="0"
      />
      <button
        class="sc-gjcoXW eGPGfw"
        type="button"
      >
        +
      </button>
    </div>
    <button
      class="sc-gjcoXW eGPGfw"
      type="submit"
    >
      Add to Cart
    </button>
  </form>
</div>
`;
